api/v1alpha1/k8ssandracluster_types.go

```diff
@@ -76,42 +76,77 @@ type K8ssandraClusterList struct {
 	Items           []K8ssandraCluster `json:"items"`
 }
 
-type Cassandra struct {
+type CassandraClusterTemplate struct {
+	// Cluster is the name of the cluster. This corresponds to cluster_name in
+	// cassandra.yaml.
+	// +kubebuilder:validation:MinLength=2
 	Cluster string `json:"cluster,omitempty"`
 
-	Datacenters []CassandraDatacenterTemplateSpec `json:"datacenters,omitempty"`
+	// ServerImage is the image for the cassandra container. Note that this should be a
+	// management-api image. If left empty the operator will choose a default image based
+	// on ServerVersion.
+	ServerImage string `json:"serverImage,omitempty"`
+
+	// ServerVersion is the Cassandra version.
+	// +kubebuilder:validation:Pattern=(3\.11\.\d+)|(4\.0\.\d+)
+	ServerVersion string `json:"serverVersion,omitempty"`
+
+	// Resources is the cpu and memory resources for the cassandra container.
+	Resources *corev1.ResourceRequirements `json:"resources,omitempty"`
+
+	// SystemLoggerResources is the cpu and memory resources for the server-system-logger
+	// container.
+	SystemLoggerResources *corev1.ResourceRequirements `json:"systemLoggerResources,omitempty"`
+
+	// CassandraConfig is configuration settings that are applied to cassandra.yaml and
+	// jvm-options for 3.11.x or jvm-server-options for 4.x.
+	CassandraConfig *CassandraConfig `json:"config,omitempty"`
+
+	// StorageConfig is the persistent storage requirements for each Cassandra pod. This
+	// includes everything under /var/lib/cassandra, namely the commit log and data
+	// directories.
+	StorageConfig *cassdcapi.StorageConfig `json:"storageConfig,omitempty"`
+
+	Racks []cassdcapi.Rack `json:"racks,omitempty"`
```

Comment:
StorageConfig is pointer, this isn't a pointer (and both are +optional?). I'd like to see consistency to which one we use.

Reply:
>StorageConfig is pointer, this isn't a pointer

I am confused by this statement.  The StorageConfig fields in both the CassandraClusterTemplate and the CassandraDatacenterTemplate are pointer so we can easily determine if they are unset.

Reply:
Usage point of view. Now whoever uses the types has both value and references in their code instead of consistently getting pointers for example ([]*cassdcapi.Rack).

Reply:
Oh, are you referring to `Racks` _not_ being a pointer?

Reply:
For slices, strings, and maps, it is easy to determine if the value is _empty_. For structs and primitive types a pointer makes it possible to distinguish whether the property has been set.